src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributeList.cs

```diff
@@ -2,73 +2,52 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
 
 namespace Microsoft.AspNetCore.Razor.TagHelpers
 {
     /// <summary>
-    /// A read-only collection of <typeparamref name="TAttribute"/>s.
+    /// A read-only collection of <typeparamref name="TagHelperAttribute"/>s.
     /// </summary>
-    /// <typeparam name="TAttribute">
-    /// The type of <see cref="IReadOnlyTagHelperAttribute"/>s in the collection.
+    /// <typeparam name="TagHelperAttribute">
+    /// The type of <see cref="TagHelperAttribute"/>s in the collection.
     /// </typeparam>
-    public class ReadOnlyTagHelperAttributeList<TAttribute> : IReadOnlyList<TAttribute>
-        where TAttribute : IReadOnlyTagHelperAttribute
+    public class ReadOnlyTagHelperAttributeList : ReadOnlyCollection<TagHelperAttribute>
```

Comment:
Agree this shortens the class. However it prevents optimizations such as not creating the backing `List<T>` unless a value is written into the read/write subclass or this is instantiated w/ a non-empty collection. Suggest holding off on this particular part of the PR until we have our offline conversation on #683.